File: src/java/_2300_Successful_Pairs_of_Spells_and_Potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int i = 0;
        int[] ans = new int[spells.length];

        Arrays.sort(potions);

        for(int spell : spells){
            int l = 0, r = potions.length - 1;

            while(l <= r){
                int m = l + (r - l) / 2;
                long product = (long)spell * potions[m];

                if(product < success)
                    l = m + 1;
                else
                    r = m - 1;
            }

            ans[i++] = potions.length - l;
        }

        return ans;
    }
}

File: src/python/_2300_Successful_Pairs_of_Spells_and_Potions.py
class Solution:
    def successfulPairs(
        self, spells: List[int], potions: List[int], success: int
    ) -> List[int]:
        ans = []
        potions.sort()

        for spell in spells:
            left = bisect_left(potions, ceil(success / spell))
            ans.append(len(potions) - left)

        return ans
